packages/core/src/SessionEventEmitter.ts

```diff
@@ -0,0 +1,256 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+import { EventEmitter } from "events";
+import { EVENTS } from "./constant";
+
+export interface ISessionEventEmitter {
+  events: SessionEventEmitter;
+}
+
+export class SessionEventEmitter extends EventEmitter {
+  constructor() {
+    super();
+  }
+  // The overloads should not be line-separated.
+  /* eslint-disable lines-between-class-members */
+  /**
+   * Register a listener called on successful login.
+   * @param eventName The login event name
+   * @param listener The callback called on successful login
+   */
+  on(eventName: typeof EVENTS.LOGIN, listener: () => void): this;
+  /**
+   * Register a listener called on successful logout.
+   * @param eventName The logout event name.
+   * @param listener The callback called on successful logout.
+   */
+  on(eventName: typeof EVENTS.LOGOUT, listener: () => void): this;
+  /**
+   * Register a listener called on session expiration.
+   * @param eventName The session expiration event name.
+   * @param listener The callback called on session expiration.
+   */
+  on(eventName: typeof EVENTS.SESSION_EXPIRED, listener: () => void): this;
+  /**
+   * Register a listener called on session restoration after a silent login.
+   * @param eventName The session restoration event name.
+   * @param listener The callback called on successful session restore.
+   */
+  on(
+    eventName: typeof EVENTS.SESSION_RESTORED,
+    listener: (currentUrl: string) => unknown
+  ): this;
+  /**
+   * Register a listener called on error, with an error identifier and description.
+   * @param eventName The error event name.
+   * @param listener The callback called on error.
+   */
+  on(
+    eventName: typeof EVENTS.ERROR,
+    listener: (
+      error: string | null,
+      errorDescription?: string | null
+    ) => unknown
+  ): this;
+  /**
+   * Register a listener called on session extension.
+   * @param eventName The session extension event name.
+   * @param listener The callback called on session extension.
+   */
+  on(
+    eventName: typeof EVENTS.SESSION_EXTENDED,
+    listener: (expiresIn: number) => void
+  ): this;
+  /**
+   * Register a listener called when a timeout is set for a session event with
+   * the timeout handle.
+   * @param eventName The timeout set event name.
+   * @param listener The callback called when setting a timeout.
+   */
+  on(
+    eventName: typeof EVENTS.TIMEOUT_SET,
+    listener: (timeoutHandle: number) => void
+  ): this;
```

Comment:
Could the `timeoutHandle` name be more desriptive here? Perhaps `millisecondsToTimeout`?

Reply:
I'll change this to `timeoutId`, as per https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#return_value